### PR TITLE
readpass handle EOF and resetsigs

### PR DESCRIPTION
**Pull request by gperciva**
This is a revised version of the readpass changes from #34 .

I'm not certain how we want to handle commit authorship.  It's not important to me, so I propose that once the code is accepted, @ijackson rebases the branch (using the "rebase fix" to combine all my commits into his, and adjusting the commit message if necessary to match the new code).

### CODE REVIEW

File: libcperciva/util/readpass.c [hunk 1]
```diff
@@ -170,6 +183,8 @@ readpass(char ** passwd, const char * prompt,
 	/* Reset terminal settings if necessary. */
 	if (usingtty)
 		tcsetattr(fileno(readfrom), TCSAFLUSH, &term_old);
+	resetsigs(savedsa);
+
```
Comment (cperciva) [inline libcperciva/util/readpass.c]: Shouldn't this be at `err2`?

File: libcperciva/util/readpass.c [hunk 2]
```diff
@@ -38,10 +38,11 @@ handle(int sig)
 	gotsig[sig] = 1;
 }
 
+/* Restore old signals and re-issue intercepted signals. */
 static void
 resetsigs(struct sigaction savedsa[NSIGS])
 {
-	int i;
+	size_t i;
```
Comment (gperciva) [inline libcperciva/util/readpass.c]: Not 100% certain you want this, but it seems to match the rest of the code better.

**Comment by gperciva** [comment]
updated PR; yes, `resetsigs()` belonged in `err2`.

**Comment by gperciva** [comment]
This PR contains all the code from #34 , with the exception of https://github.com/ijackson/scrypt/commit/dda36e2a280faab52a288d5b48b32582063a855d "readpass: Always clean up on success path even if signaled", of which @cperciva said:
> I'm not convinced. We needed to capture signals so that we could ensure that we reset the terminal state if we were killed; but if the caller sets up signals to siglongjmp out of readpass then they get what they deserve.

I'm going to close the old PR so that we don't have two PRs for the same code.

**Comment by gperciva** [comment]
Updated PR with rebased and minor rewritten commit messages, with permission from @ijackson.

**Comment by cperciva** [comment]
Thanks!  Sorry it took so long for me to merge this.

**Pull request merged**